lib/compress/zstd_compress.c

```diff
@@ -1814,17 +1818,35 @@ static size_t ZSTD_compressBlock_internal(ZSTD_CCtx* zc,
     {   U32 const extDict = ZSTD_window_hasExtDict(ms->window);
         size_t lastLLSize;
         { int i; for (i = 0; i < ZSTD_REP_NUM; ++i) zc->blockState.nextCBlock->rep[i] = zc->blockState.prevCBlock->rep[i]; }
-        if (zc->appliedParams.ldmParams.enableLdm) {
-            size_t const nbSeq =
-                ZSTD_ldm_generateSequences(&zc->ldmState, zc->ldmSequences,
-                                           &zc->appliedParams.ldmParams,
-                                           src, srcSize, extDict);
+        if (zc->externSeqStore.pos < zc->externSeqStore.end) {
+            assert(!zc->appliedParams.ldmParams.enableLdm);
             lastLLSize =
-                ZSTD_ldm_blockCompress(zc->ldmSequences, nbSeq,
+                ZSTD_ldm_blockCompress(&zc->externSeqStore,
                                        ms, &zc->seqStore,
                                        zc->blockState.nextCBlock->rep,
                                        &zc->appliedParams.cParams,
                                        src, srcSize, extDict);
+            assert(zc->externSeqStore.pos <= zc->externSeqStore.end);
+        } else if (zc->appliedParams.ldmParams.enableLdm) {
+            rawSeqStore_t ldmSeqStore;
+
+            ldmSeqStore.seq = zc->ldmSequences;
+            ldmSeqStore.end = zc->maxNbLdmSequences;
+            ldmSeqStore.pos = 0;
+
+            CHECK_F(ZSTD_ldm_generateSequences(&zc->ldmState, &ldmSeqStore,
+                                               &zc->appliedParams.ldmParams,
+                                               src, srcSize));
+            /* We generated ldmSeqStore.pos sequences */
+            ldmSeqStore.end = ldmSeqStore.pos;
```

Comment:
The meaning of `end` and `pos` is changed
depending on being passed to `ZSTD_ldm_generateSequences()` or `ZSTD_ldm_blockCompress()`.

While I'm sure you properly control what's going on today,
I'm worried that some future contributor (which can be any one of us in the not-so-distance future), might be confused by such change of meaning half-way in the process.

I would really prefer `capacity`, `end`, and `pos` to be completely separated.
`ZSTD_ldm_generateSequences()` would update `end` by creating new sequences, and ensure it never goes beyond `capacity`,
`ZSTD_ldm_blockCompress()` would consume sequences by updating `pos`, ensuring it never goes beyond `end`.